django/cantusdb_project/main_app/views/views.py

```diff
@@ -654,6 +623,141 @@ def build_json_cid_dictionary(chant, request) -> dict:
     return dictionary
 
 
+def record_exists(rec_type: BaseModel, pk: int) -> bool:
+    """Determines whether record of specific type (chant, source, sequence, article) exists for a given pk
+
+    Args:
+        rec_type (BaseModel): Which model to check to see if an object of that type exists
+        pk (int): The ID of the object being checked for.
+
+    Returns:
+        bool: True if an object of the specified model with the specified ID exists, False otherwise.
+    """
+    try:
+        rec_type.objects.get(id=pk)
+        return True
+    except rec_type.DoesNotExist:
+        return False
+
+
+def get_user_id_from_old_indexer_id(pk: int) -> Optional[int]:
+    """
+    Finds the matching User ID in NewCantus for an Indexer ID in OldCantus.
+    This is stored in the User table's old_indexer_id column.
+    This is necessary because indexers were originally stored in the general Node
+    table in OldCantus, but are now represented as users in NewCantus.
+
+    Args:
+        pk (int): the ID of an indexer in OldCantus
+
+    Returns:
+        Optional int: the ID of the corresponding User in NewCantus
+    """
+    User = get_user_model()
+    try:
+        result = User.objects.get(old_indexer_id=pk)
+        return result.id
+    except User.DoesNotExist:
+        return None
+
+
+def check_for_unpublished(item: Union[Chant, Sequence, Source]) -> None:
+    """Raises an Http404 exception if item is unpublished
+
+    Args:
+        item (Chant, Sequence, or Source): An item to check whether it is published or not
+
+    Raises:
+        Http404 if the item is a source and it's unpublished,
+            or if it's a chant/sequence and its source is unpublished
+
+    Returns:
+        None
+    """
+    if isinstance(item, Source):
+        if not item.published:
+            raise Http404()
+    if isinstance(item, Chant) or isinstance(item, Sequence):
+        if not item.source.published:
+            raise Http404()
+
+
+NODE_TYPES_AND_VIEWS = [
+    (Chant, "chant-detail"),
+    (Source, "source-detail"),
+    (Sequence, "sequence-detail"),
+    (Article, "article-detail"),
+]
+
+
+# all IDs above this value are created in NewCantus and thus could have conflicts between types.
+# when data is migrated from OldCantus to NewCantus, (unpublished) dummy objects are created
+# in the database to ensure that all newly created objects have IDs above this number.
+NODE_ID_CUTOFF = 1_000_000
+
+
+def json_node_export(request, id: int) -> JsonResponse:
+    """
+    returns all fields of the chant/sequence/source/indexer with the specified `id`
+    """
+
+    # all IDs above this value are created in NewCantus and thus could have conflicts between types.
+    # when data is migrated from OldCantus to NewCantus, (unpublished) dummy objects are created
```

Comment:
Are these unpublished dummy objects kept around or deleted?

You can also just run a single SQL command once to set the starting ID in the new database... `ALTER SEQUENCE "the_table_sequence_id" RESTART WITH 100000000;`.

Reply:
They can be kept around, or deleted - they're unpublished (so only visible to logged-in users), and basically serve to do what your command does. They have a message saying they can be deleted once a source/chant/sequence with an ID greater than 1000000 has been created, so they shouldn't cause too much confusion.

But, @lucasmarchd01, based on this, while doing the final data sync, we can probably just remove the bit about creating the dummy objects if it makes things simpler and then run this command instead (n.b. with the right number of zeroes).